[WIP] Reconstruction Miniapp

cjvogl opened the pull request:
TODO list

- [x] add quantitative evaluation of reconstruction (e.g., discrete $L^2$ and/or $L^\inf$ error) to terminal output
- [x] refinement of mesh via command line arguments
- [x] replace static implementation of 4th row of $A$ (enforcing $\frac{d^2u}{dxdy} = 0$) with integrator implementation

cjvogl committed:
initial reconstruction miniapp

cjvogl commented on miniapps/tools/reconstruction.cpp:
```diff
@@ -0,0 +1,178 @@
+// Note current only 2D reconstruction of L^2_1 (piecewise-linear) field from
+// L^2_0 (piecewise-constant) field is supported. The reconstruction is done
+// by enforcing
+//   1) element average: (u, psi_hat)_E = ({u_hat}, psi_hat)_E
```
```suggestion
//   1) element average: (u, psi_hat)_E = (u_hat, psi_hat)_E
```

cjvogl commented on miniapps/tools/reconstruction.cpp:
```diff
@@ -0,0 +1,178 @@
+      Vector b(dst_element.GetDof());
+      DenseMatrix Arow;
+
+      // enforce (u, psi_hat)_E = ({u_hat}, psi_hat)_E
```
```suggestion
      // enforce (u, psi_hat)_E = (u_hat, psi_hat)_E
```

cjvogl committed:
Correcting bracket notation

homeomorfismo committed:
Make ASTYLE_BIN optional if set

homeomorfismo committed:
Force style

homeomorfismo committed:
Add serial and parallel refinement

homeomorfismo committed:
Merge branch 'master' of github.com:mfem/mfem into reconstruction

cjvogl commented on miniapps/tools/reconstruction.cpp:
```diff
@@ -152,27 +189,27 @@ int main(int argc, char* argv[])
+      // TODO: quantitatively compare (e.g. L2 norm) original & reconstruction
 
-   Mpi::Finalize();
+      Mpi::Finalize();
```
I dont think this two lines should have had the indent changed?

homeomorfismo replied:
I checked the git-blame and it seems it could be an `astyle` thing. I wrapped the else block so I can keep the git hook active haha

homeomorfismo committed:
L2 Norm error

homeomorfismo committed:
Exposing NCMesh::FindNeighbors

homeomorfismo committed:
CSV output (for bash scripts)

homeomorfismo committed:
Least squares driver

homeomorfismo committed:
Merge branch 'master' of github.com:mfem/mfem into reconstruction

homeomorfismo committed:
Add visualization, almost working

homeomorfismo committed:
Error estimation, and averages of reconstruction; scaling issue

homeomorfismo committed:
Fix scaling; need fix orientation

homeomorfismo committed:
Cleaner notes and integrator

homeomorfismo committed:
Add convergence rates snippet

homeomorfismo committed:
Better comments; use references and pointers

homeomorfismo committed:
Add preserve volumes

homeomorfismo committed:
Correct CLS

homeomorfismo committed:
Dont use CG for saddle point

homeomorfismo committed:
Correct avg error computation

homeomorfismo committed:
Clarify error message and CLS

homeomorfismo committed:
Expose Neighbor function

homeomorfismo committed:
Temporary logic for HO Rec

homeomorfismo committed:
Add r/atol, add user-given regulatization

homeomorfismo committed:
Working draft of MbdHessianInt

homeomorfismo committed:
Check LS: Issues with tolerances...

homeomorfismo committed:
Add draft generic solver (using MINRES)

homeomorfismo committed:
Partial progress Solver

homeomorfismo committed:
Merge branch 'master' of github.com:mfem/mfem into reconstruction

homeomorfismo committed:
Remove integrator comment that confuses the big machine

homeomorfismo committed:
Add L1-Jacobi preconditoner for AtA + misc

homeomorfismo committed:
Extrude FindNeighbors logic; add better conditional

homeomorfismo committed:
Suport to direct solvers

homeomorfismo committed:
Add more solvers

homeomorfismo committed:
Copyright LLNL

homeomorfismo committed:
Consistency is good

homeomorfismo committed:
Clean up attempted bidirectional integrator misc

homeomorfismo committed:
Not-so-drop-in replacement for Hessian

homeomorfismo committed:
Consistency among drivers

homeomorfismo committed:
Add visualization options

homeomorfismo committed:
Replace dev with todo in Doxygen comments

homeomorfismo committed:
Stronger checks on Newton solver

homeomorfismo committed:
Add print level

homeomorfismo committed:
Better names; Attempt to fix CLS

homeomorfismo committed:
Get faces averages on new driver

homeomorfismo committed:
Fix formatting, replace some prints, rename vars

homeomorfismo committed:
Add complete out-of-element matrix (L2LS)

homeomorfismo committed:
Does the mesh have enough elements?

homeomorfismo committed:
Both residual and solution (still a misnomer)

cjvogl committed:
update build system so new miniapps are considered parallel

cjvogl committed:
marked least-squares and flux-reconstruction drivers as serial for now

homeomorfismo committed:
Make params structs

homeomorfismo committed:
Change smallness criteria

homeomorfismo committed:
L2Reconstruction compiles

homeomorfismo committed:
From more options to a struct

homeomorfismo committed:
Extude main function

homeomorfismo committed:
Minor fixes and docstrings

homeomorfismo committed:
Remove main logic to outside function

homeomorfismo committed:
Merge branch 'master' of github.com:mfem/mfem into reconstruction

homeomorfismo committed:
Scope GLVis, MPI solvers

homeomorfismo committed:
Preserve volumes in AverageReconstruction exposed

homeomorfismo committed:
Get ParFES, remove static_cast

homeomorfismo committed:
Attempt to fix preserve volumes

homeomorfismo committed:
remove variant in favor of structs

homeomorfismo committed:
Typo and print lvl

homeomorfismo committed:
Generalize the source order for L2 rec

homeomorfismo committed:
Update driver

homeomorfismo committed:
Move flux-reconstruction into least-squares driver

homeomorfismo committed:
Remove flux-reconstruction driver (in LS now)

homeomorfismo committed:
Add Face-avg option

homeomorfismo committed:
Write most of the body for DFR-like

homeomorfismo committed:
make style

homeomorfismo committed:
Update CMake file accordingly

homeomorfismo committed:
Almost complete DFR; requires extrude IR

homeomorfismo committed:
Refactor code; get common HO IR

homeomorfismo committed:
Barely working method

homeomorfismo committed:
Dox strings fix

homeomorfismo committed:
Force average in boundary terms (assuming null brd)

homeomorfismo committed:
Correct namespace

homeomorfismo committed:
Suggested LS implementation

homeomorfismo committed:
Fix odd error on mesh

homeomorfismo committed:
Modify constrained implementation for AverageReconstruction

homeomorfismo committed:
Use OperatorPtr in LSSolver

homeomorfismo committed:
Use SchurConstrainedSolver

homeomorfismo committed:
Small missing safeguard

homeomorfismo committed:
Rewrite AverageReconstruction

homeomorfismo committed:
Other minor clean-ups

homeomorfismo committed:
Clean up L2Reconstruction

homeomorfismo committed:
Correct some aspects of strong FaceReconstruction

homeomorfismo committed:
L2-constrained working

homeomorfismo committed:
Gitignore, shadowing var

homeomorfismo committed:
Simplify with GetElementDofValues

homeomorfismo committed:
Partial progress on WFR and BVR

homeomorfismo committed:
Set size bug

homeomorfismo committed:
Add optional mass matrix regularization

homeomorfismo committed:
Barebones idea different regularizations

homeomorfismo committed:
Add different regulatization terms

homeomorfismo committed:
Merge branch 'master' of github.com:mfem/mfem into reconstruction

homeomorfismo committed:
More examples on run-time

homeomorfismo committed:
Remove some unused vars

homeomorfismo committed:
Return IntRule makes more sense

homeomorfismo committed:
Draft Average integrator

homeomorfismo committed:
Use GetNDofs instead

homeomorfismo committed:
Minor correction

homeomorfismo committed:
Weak FR reconstruction idea

homeomorfismo committed:
Temporal removal of CG Piola checks

homeomorfismo committed:
Fix FaceReconstruction: provide elem idx and choose right transform

homeomorfismo committed:
Use references for fes, raw pointers for trans

homeomorfismo committed:
Resize one vector

homeomorfismo committed:
Misc cleanup

homeomorfismo committed:
Merge branch 'master' of github.com:mfem/mfem into reconstruction

homeomorfismo committed:
Prefer references

homeomorfismo committed:
Propagate to WFR

homeomorfismo committed:
Add missing sign

cjvogl commented on mesh/ncmesh.hpp:
```diff
@@ -1025,6 +1025,7 @@ class NCMesh
                          Array<int> *neighbors, /* append */
                          Array<char> *neighbor_set = NULL);
 
+public:
```
I'm wondering why `FindNeighbors` and `NeighborExpand` weren't already public.  Do they modify member variables?  If not, perhaps make them `const` member functions would alleviate any concerns with making them `public`

cjvogl commented on miniapps/tools/CMakeLists.txt:
```diff
@@ -53,16 +53,22 @@ add_mfem_miniapp(tmop-metric-magnitude
   ${MFEM_MINIAPPS_COMMON_HEADERS}
   LIBRARIES mfem-common)
 
+add_mfem_miniapp(least-squares
+  MAIN least-squares.cpp LIBRARIES mfem)
```
Will `least-squares` work if `MFEM_USE_MPI=OFF`?  I suspect there may be issues because the `Par`-prefixed classes are used, but I could be wrong

cjvogl commented on miniapps/tools/least-squares.cpp:
```diff
@@ -0,0 +1,1720 @@
+{
+
+/// Enumerator for different function examples
+enum ExamplesType
+{
+   plane,
+   sine,
+   exp_sine,
+   num_examples  // last
+};
```
I would consider partitioning this file in general into a header file with these enum, class, and functions definitions.

cjvogl commented on miniapps/tools/least-squares.cpp:
```diff
@@ -0,0 +1,1720 @@
+   switch (params.rec)
+   {
+      case norm:
+         L2Reconstruction(*solver.get(), u_src, u_dst, mesh, params.newton,
+                          params.reg_type, params.reg, params.preserve_volumes);
```
Note this can also be applied in the other reconstruction methods.

cjvogl commented on miniapps/tools/least-squares.cpp:
```diff
@@ -0,0 +1,1720 @@
+/** Integrator of the DPG form $\langle v, \lbrace w \rbrace \rangle$ where
+ *  $\lbrace \cdot \rbrace$ represents the average of the traces of the argument.
+ */
+class TraceAverageIntegrator : public BilinearFormIntegrator
```
Similar to a comment about the `AsymmetricMassIntegrator`, does this class need to inherit from `BilinearFormIntegrator`?  More specifically, is a `TraceAverageIntegrator` ever used as a `BilinearFormIntegrator`?  If not, please remove this inheritance.  For what it's worth, this is known as preferring composition ("has a") over inheritance ("is a").

cjvogl commented on miniapps/tools/least-squares.cpp:
```diff
@@ -0,0 +1,1720 @@
+/// @name Auxiliary functions
+///@{
+
+/// @brief Dense small least squares solver
```
Also, please note what the method is that is used to solved the problem... something like 
"solves the least squares problem $x = argmin_y \|Ay - b\|_2^2$ by enforcing the first-order optimality condition $A^TA x - A^T b = 0$

homeomorfismo commented on makefile:
```diff
@@ -791,7 +791,7 @@ status info:
 	@true
 
-ASTYLE_BIN = astyle
+ASTYLE_BIN ?= astyle
```
TLDR: Yes. 

As far as I know, the make rules of the astyle repo create the binaries in an odd directory tree and the binary is usually named `Astyle`. To avoid cluttering someone's `PATH`, the binary could be provided as an env variable. I mainly did this to run the git hooks comfortably, but the commit can be reverted too. :)

cjvogl commented on miniapps/tools/least-squares.cpp:
```diff
@@ -0,0 +1,1720 @@
+///@{
+
+/// @brief Local-averages-based reconstruction
+void AverageReconstruction(Solver& solver,
+                           ParGridFunction& src,
+                           ParGridFunction& dst,
+                           ParMesh& mesh,
+                           IterativeSolverParams& newton,
+                           RegularizationType reg_type = direct,
+                           real_t reg = 0.0,
+                           int print_level = -1,
+                           bool preserve_volumes = false)
```
I believe the `ParMesh` object can be obtain through one of the `ParGridFunction` arguments, so I would remove it from the function signature.

Additionally on a first pass read, I believe `print_level` is used more as a boolean flag than an integer print level... if that is the case, I would change the function signature to use a boolean flag.

cjvogl commented on miniapps/tools/least-squares.cpp:
```diff
@@ -0,0 +1,1720 @@
+/// @brief Dense small least squares solver, with constrains @a C with value @a c
+/// but the constraint is one-dimensional, and passed as a vector. Does not return
+/// the multiplier.
+void LSSolver(Solver& solver,
```
To disambiguate the `solver` argument from the function which has Solver in the name, I suggest renaming to `linear_solver` (applies to all other use of `solver` as well)

cjvogl commented on miniapps/tools/least-squares.cpp:
```diff
@@ -0,0 +1,1720 @@
+   }
+}
+
+/// @brief An out-of-element, L2-based reconstruction
```
Please expand the comments here to describe the optimization problem being solved and the method for solving it (see comments on LSSolver)

cjvogl commented on miniapps/tools/least-squares.cpp:
```diff
@@ -0,0 +1,1720 @@
+    * later on. DenseMatrix does not have this method
+    * implemented.
+    */
+   std::shared_ptr<Solver> solver(nullptr);
```
Technically, I think the trivial constructor already sets the object to `nullptr`, so you can replace with `std::shared_ptr<Solver> solver`... although I would replace `solver` with the more descriptive `linear_solver`

cjvogl commented on miniapps/tools/least-squares.cpp:
```diff
@@ -0,0 +1,1720 @@
+         case 0:
+            print_level.All();
+            break;
+         case 1:
+            print_level.FirstAndLast();
+            print_level.Warnings();
+         case 2:
+            print_level.Errors();
+            break;
```
Missing a break statement as part of `case 1`?

cjvogl commented on miniapps/tools/least-squares.cpp:
```diff
@@ -0,0 +1,1720 @@
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
```
Please add some "sample runs" that contain an example of each of the methods you've implemented.

homeomorfismo committed:
Undo references for tranformations

- `FiniteElementSpace::GetElementTransformation(int)` wraps
  `Mesh::GetElementTransformation(int)`, and returns a reference
  owned by the mesh.
- `FiniteElementSpace::GetElementTransformation(int, T*)` requires
  `ElementTransformation` to be allocated.
- `FaceElementTransformations` copy assigment operator is implicitly
  deleted.
- Ideally, we would use `Mesh::GetFaceElementTransformations(int, T1&,
  T2&, T3&)`, but I avoided dealing with more variables.

cjvogl committed:
minor clean up of visualization code

cjvogl committed:
started to consolidate the reconstruction and least-squares files

cjvogl committed:
minor cleanup of comments

skang67 committed:
Added LOR reconstruction option and clarified serial nature of current methods. (#5270)

Implementation complements of @skang67

cjvogl committed:
Merge branch 'master' into reconstruction

skang67 committed:
Parallel LOR reconstruction miniapp. (#5298)

Co-authored-by: Chris Vogl <vogl2@llnl.gov>

cjvogl committed:
moved reconstruction apps to new samr miniapps directory

cjvogl commented:
Closing this PR in lieu of #5362 that will include all the reconstruction work here alongside the SAMRAI-MFEM coupling work.
